### PR TITLE
feat: ci-publish-npm action

### PR DESCRIPTION
**Pull request by erikburt**
As part of [RE-2033](https://smartcontract-it.atlassian.net/browse/RE-2033) - create a reusable npm publishing action based off [smartcontractkit/ccip/.github/workflows/solidity](https://github.com/smartcontractkit/ccip/blob/7d332c917a190a5ef89e279213f72968efc4fe32/.github/workflows/solidity.yml#L162-L221).

Will be used initially in https://github.com/smartcontractkit/chainlink-contracts-deprecated/pull/2

Testing:
- https://github.com/smartcontractkit/chainlink-contracts-deprecated/actions/runs/7147133338/job/19466124711

<details>
  <summary><i>Test run output</i></summary>

```
Run smartcontractkit/.github/actions/ci-publish-npm@feat/ci-publish-npm
Run echo "//registry.npmjs.org/:_authToken=${NODE_AUTH_TOKEN}" >> ~/.npmrc
Run pnpm publish-test --publish-branch 'feat/publishing-follow-up' --no-git-checks
> @chainlink/contracts-deprecated@0.1.0 publish-test /home/runner/work/chainlink-contracts-deprecated/chainlink-contracts-deprecated/contracts
> ./scripts/echo-test.sh "--publish-branch" "feat/publishing-follow-up" "--no-git-checks"
--publish-branch
feat/publishing-follow-up
--no-git-checks

Run smartcontractkit/.github/actions/ci-publish-npm@feat/ci-publish-npm
Run echo "//registry.npmjs.org/:_authToken=${NODE_AUTH_TOKEN}" >> ~/.npmrc
Run ./contracts/scripts/echo-test.sh --publish-branch 'feat/publishing-follow-up' --no-git-checks
--publish-branch
feat/publishing-follow-up
--no-git-checks

Run cat  ~/.npmrc
//registry.npmjs.org/:_authToken=fake-test-token
registry=https://registry.npmjs.org/
//registry.npmjs.org/:_authToken=fake-test-token
registry=https://registry.npmjs.org/
```
</details>

### CODE REVIEW

File: actions/ci-publish-npm/action.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    required: false
+    default: '.'
+  publish-command:
+    description: "The package.json defined script used to publish. Will run as `pnpm <publish-command>`"
+    required: true
+  npm-token:
+    description: The NPM token with publish access
+    required: true
+  github-token:
+    description: The GITHUB_TOKEN for the workflow
```
Comment (momentmaker) [inline actions/ci-publish-npm/action.yml:16]: should we say something like what level of perm it requires in order to run the `action-gh-release`?

**Review by momentmaker — APPROVED** [review]

File: actions/ci-publish-npm/action.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      env:
+        NODE_AUTH_TOKEN: ${{ inputs.npm-token }}
+      run: |
+        echo "//registry.npmjs.org/:_authToken=${NODE_AUTH_TOKEN}" >> ~/.npmrc
+        echo "registry=https://registry.npmjs.org/" >> ~/.npmrc
+
+    - name: Publish to NPM
+      shell: bash
+      working-directory: ${{ inputs.package-json-directory }}
+      run: pnpm ${{ inputs.publish-command }}
```
Comment (erikburt) [inline actions/ci-publish-npm/action.yml:50]: Should this allow for an arbitrary command or be restricted to a script declared in the `package.json` (current).

Comment (momentmaker) [reply actions/ci-publish-npm/action.yml:50]: ahh good call. I'd imagine we should go with the latter option - maybe a check prior to this step to check whether `publish-command` input is in the `scripts` object?

Comment (erikburt) [reply actions/ci-publish-npm/action.yml:50]: Discussed in a meeting - will be switching this action to allow for arbitrary commands. The indirection is not worthwhile, and the "risk" of arbitrary command excution is non-existent as calling workflows/actions could execute the same command anyway outside of the scope of this action.

**Pull request merged**